camayoc/tests/qcs/cli/test_credentials.py

```diff
@@ -0,0 +1,751 @@
+        '    "username": "{}"\r\n'
+        '}}\r\n'
+        .format(name, MASKED_PASSWORD_OUTPUT, username)
+    ) == 0
+    assert qpc_cred_show.expect(pexpect.EOF) == 0
+    qpc_cred_show.close()
+    assert qpc_cred_show.exitstatus == 0
+
+
+def test_add_with_username_password_sudo_password(
+        isolated_filesystem, qpc_server_config):
+    """Add an auth with username, password and sudo password.
+
+    :id: df0d61d5-363f-400a-961c-04146f6089e1
+    :description: Add an auth entry providing the ``--name``, ``--username``,
+        ``--pasword`` and ``--sudo-password`` options.
+    :steps: Run ``qpc cred add --name <name> --username <username> --password
+        --sudo-password``
+    :expectedresults: A new auth entry is created with the data provided as
+        input.
+    """
+    name = utils.uuid4()
+    username = utils.uuid4()
+    cred_add(
+        {
+            'name': name,
+            'username': username,
+            'password': None,
+            'sudo-password': None,
+        },
+        [
+            (CONNECTION_PASSWORD_INPUT, utils.uuid4()),
+            (SUDO_PASSWORD_INPUT, utils.uuid4()),
+        ],
+    )
+
+    qpc_cred_show = pexpect.spawn(
+        'qpc cred show --name={}'.format(name)
+    )
+    qpc_cred_show.logfile = BytesIO()
+    assert qpc_cred_show.expect(
+        '{{\r\n'
+        '    "cred_type": "network",\r\n'
+        '    "id": \\d+,\r\n'
+        '    "name": "{}",\r\n'
+        '    "password": "{}",\r\n'
+        '    "sudo_password": "{}",\r\n'
+        '    "username": "{}"\r\n'
+        '}}\r\n'
+        .format(name, MASKED_PASSWORD_OUTPUT, MASKED_PASSWORD_OUTPUT, username)
+    ) == 0
+    assert qpc_cred_show.expect(pexpect.EOF) == 0
+    qpc_cred_show.close()
+    assert qpc_cred_show.exitstatus == 0
+
+
+def test_add_with_username_sshkeyfile(isolated_filesystem, qpc_server_config):
+    """Add an auth with username and sshkeyfile.
+
+    :id: 9bfb16a2-5a10-4a01-9f9d-b29445c1f4bf
+    :description: Add an auth entry providing the ``--name``, ``--username``
+        and ``--sshkeyfile`` options.
+    :steps: Run ``qpc cred add --name <name> --username <username> --sshkeyfile
+        <sshkeyfile>``
+    :expectedresults: A new auth entry is created with the data provided as
+        input.
+    """
+    name = utils.uuid4()
+    username = utils.uuid4()
+    sshkeyfile = Path(utils.uuid4())
+    sshkeyfile.touch()
+    cred_add({
+        'name': name,
+        'username': username,
+        'sshkeyfile': sshkeyfile.name,
+    })
+
+    qpc_cred_show = pexpect.spawn(
+        'qpc cred show --name={}'.format(name)
+    )
+    qpc_cred_show.logfile = BytesIO()
+    assert qpc_cred_show.expect(
+        '{{\r\n'
+        '    "cred_type": "network",\r\n'
+        '    "id": \\d+,\r\n'
+        '    "name": "{}",\r\n'
+        '    "ssh_keyfile": "{}",\r\n'
+        '    "username": "{}"\r\n'
+        '}}\r\n'
+        .format(name, sshkeyfile.resolve(), username)
```

Comment:
Perhaps to avoid repeating this segment of code with the expected output, you could have a little helper function that does the string interpolation for you and so you will just have to update that one place if this output changes.

Reply:
This is a good suggestion, will handle it.

Reply:
Taking a look in order to implement a function to help generating the output there are some requirements:

1) The fields should have the same order
2) Not all outputs are the same. It will depend on the information that the credential has, for example if sudo_password is not set then it won't be present on the output.

By creating a function we will need to find a way to pass the information to generate the output and the first option that comes to mind is to use dictionaries. The problem is that the order is not guaranteed across all minor versions (they are ordered on python 3.5+). So in order to have this working across different versions we will need to use either an OrderedDict  or assume Camayoc will be running on Python 3.5+ (which should be the case on our dev and jenkins machines).

Are you OK on assuming we will be running Camayoc on Python 3.5+ and use simple dicts or shall we enforce the order with OrderedDict?

Reply:
Hmm gut reaction is use OrderedDict

Reply:
Using an OrderedDict requires us to sort the dict but it also continue to repeat some code:

```diff
diff --git a/camayoc/tests/qcs/cli/test_credentials.py b/camayoc/tests/qcs/cli/test_credentials.py
index bd0e2a5..8b41274 100644
--- a/camayoc/tests/qcs/cli/test_credentials.py
+++ b/camayoc/tests/qcs/cli/test_credentials.py
@@ -10,6 +10,7 @@
 :upstream: yes
 """
 import random
+from collections import OrderedDict
 from io import BytesIO
 from pathlib import Path
 
@@ -22,7 +23,7 @@ from camayoc.constants import (
     MASKED_PASSWORD_OUTPUT,
     SUDO_PASSWORD_INPUT,
 )
-from camayoc.tests.qcs.cli.utils import cred_add
+from camayoc.tests.qcs.cli.utils import cred_add, cred_show
 
 
 def test_add_with_username_password(isolated_filesystem, qpc_server_config):
@@ -48,23 +49,14 @@ def test_add_with_username_password(isolated_filesystem, qpc_server_config):
         ],
     )
 
-    qpc_cred_show = pexpect.spawn(
-        'qpc cred show --name={}'.format(name)
+    cred_show(
+        {'name': name},
+        OrderedDict(sorted({
+            'name': name,
+            'username': username,
+            'password': MASKED_PASSWORD_OUTPUT,
+        }.items(), key=lambda i: i[0]))
     )
-    qpc_cred_show.logfile = BytesIO()
-    assert qpc_cred_show.expect(
-        '{{\r\n'
-        '    "cred_type": "network",\r\n'
-        '    "id": \\d+,\r\n'
-        '    "name": "{}",\r\n'
-        '    "password": "{}",\r\n'
-        '    "username": "{}"\r\n'
-        '}}\r\n'
-        .format(name, MASKED_PASSWORD_OUTPUT, username)
-    ) == 0
-    assert qpc_cred_show.expect(pexpect.EOF) == 0
-    qpc_cred_show.close()
-    assert qpc_cred_show.exitstatus == 0
 
 
 def test_add_with_username_password_sudo_password(
@@ -94,24 +86,15 @@ def test_add_with_username_password_sudo_password(
         ],
     )
 
-    qpc_cred_show = pexpect.spawn(
-        'qpc cred show --name={}'.format(name)
+    cred_show(
+        {'name': name},
+        OrderedDict(sorted({
+            'name': name,
+            'password': MASKED_PASSWORD_OUTPUT,
+            'sudo_password': MASKED_PASSWORD_OUTPUT,
+            'username': username,
+        }.items(), key=lambda i: i[0]))
     )
-    qpc_cred_show.logfile = BytesIO()
-    assert qpc_cred_show.expect(
-        '{{\r\n'
-        '    "cred_type": "network",\r\n'
-        '    "id": \\d+,\r\n'
-        '    "name": "{}",\r\n'
-        '    "password": "{}",\r\n'
-        '    "sudo_password": "{}",\r\n'
-        '    "username": "{}"\r\n'
-        '}}\r\n'
-        .format(name, MASKED_PASSWORD_OUTPUT, MASKED_PASSWORD_OUTPUT, username)
-    ) == 0
-    assert qpc_cred_show.expect(pexpect.EOF) == 0
-    qpc_cred_show.close()
-    assert qpc_cred_show.exitstatus == 0
 
 
 def test_add_with_username_sshkeyfile(isolated_filesystem, qpc_server_config):
diff --git a/camayoc/tests/qcs/cli/utils.py b/camayoc/tests/qcs/cli/utils.py
index a0267a1..53f7004 100644
--- a/camayoc/tests/qcs/cli/utils.py
+++ b/camayoc/tests/qcs/cli/utils.py
@@ -32,3 +34,35 @@ def cred_add(options, inputs=None, exitstatus=0):
     assert qpc_cred_add.expect(pexpect.EOF) == 0
     qpc_cred_add.close()
     assert qpc_cred_add.exitstatus == exitstatus
+
+
+def cred_show(options, output, exitstatus=0):
+    r"""Show a credential entry.
+
+    :param options: A dictionary mapping the option names and their values.
+        Pass ``None`` for flag options.
+    :param output: An OrderedDict mapping the output fields and their values.
+        The order the fields were inserted in the OrderedDict will be the same
+        order they will be present on the output. If cred_type and/or id fields
+        are not present on the output they will be added with the value of
+        network and \d+ respectively.
+    :param exitstatus: Expected exit status code.
+    """
+    command = 'qpc cred show'
+    for key, value in options.items():
+        if value is None:
+            command += ' --{}'.format(key)
+        else:
+            command += ' --{}={}'.format(key, value)
+    qpc_cred_show = pexpect.spawn(command)
+    expected_output = '{{\r\n'
+    if 'cred_type' not in output:
+        expected_output += '    "cred_type": "network",\r\n'
+    if 'id' not in output:
+        expected_output += '    "id": \\d+,\r\n'
+    for key, value in output.items():
+        expected_output += '    "{}": "{}",\r\n'.format(key, value)
+    assert qpc_cred_show.expect(expected_output)
+    assert qpc_cred_show.expect(pexpect.EOF) == 0
+    qpc_cred_show.close()
+    assert qpc_cred_show.exitstatus == exitstatus
```

Reply:
We can have a helper function to generate the sorted ordered dict for the show output values. Or we can make Camayoc require python 3.5+ and we should have a very clean code.

To be honest I am more inclined to embrace the future and make Camayoc Python 3.5+ only :smile:

Reply:
Since we are our own best and only customers, I think enforcing a version is fine, especially if we can install the quipucords server via rpm or run in a different venv/shell. We just need to be able to run tests against quipucords running in all supported configurations.